lib/internal/quic/core.js

```diff
@@ -2590,11 +2651,64 @@ class QuicStream extends Duplex {
   }
 
   [kStreamReset](code) {
-    this[kInternalState].resetCode = code | 0;
+    // Receiving a reset from the peer indicates that it is no
+    // longer sending any data, we can safely close the readable
+    // side of the Duplex here.
+    this[kInternalState].resetCode = code;
     this.push(null);
     this.read();
   }
 
+  [kClose]() {
+    const state = this[kInternalState];
+
+    if (this.destroyed) {
+      return PromiseReject(
+        new ERR_INVALID_STATE('QuicStream is already destroyed'));
+    }
+
+    const promise = deferredClosePromise(state);
+    if (this.readable) {
+      this.push(null);
+      this.read();
+    }
+
+    if (this.writable) {
+      this.end();
+    }
+
+    return promise;
```

Comment:
I don't think you need this? Using `finished(this)` should be sufficient?

Reply:
Don't need what precisely?

Reply:
instead of `return promise` I think you could do `return finished(this)`.

Reply:
and skip closePromiseReject, closePromiseResolve etc.

Reply:
Perhaps, `finished()` does not, however, return a `Promise`, and that would make the `QuicStream` `close()` behavior diverge from `QuicSocket` and `QuicSession`, which isn't ideal. We can potentially revisit that later. Let's leave this as is for now.

Reply:
> finished() does not, however, return a Promise

There is a promisified version of it?

Reply:
Again, please add a TODO comment if we plan to revisit this later so we don't forget.

Reply:
I'll add a to-do for sure. I look into the promisified version. If it's just a wrapper impl tho I'd prefer to avoid it.